ldp/alg/algorithms.py

```diff
@@ -126,73 +126,128 @@ def gvizify(x: Any) -> str:
-            extract_answer_fn = cast(
-                Callable[[TData], Awaitable[TAnswer]], extract_answer_fn
-            )
-            answers = await asyncio.gather(*(extract_answer_fn(x) for x in sampled))
-        else:
-            extract_answer_fn = cast(Callable[[TData], TAnswer], extract_answer_fn)
-            answers = [extract_answer_fn(x) for x in sampled]
-
-        # Compute consensus: mode of the sampled answers
-        grouped_consensus[group_key] = collections.Counter(answers).most_common()
-        # NOTE: If there are multiple modes, just use the first one
-        consensus: TAnswer = grouped_consensus[group_key][0][0]
-        if ideal_answer_fn != NO_IDEAL_ANSWER_FN:
+
+    async def index(group_key: TGroupKey, group: list[TData]) -> int:
+        grouped_consensus[group_key], consensus = await evaluate_consensus(
+            group, extract_answer_fn, num_samples, rand, consensus_callback
+        )
+        if ideal_answer_fn != NO_IDEAL_ANSWER_FN:  # If we have an ideal
             # Assume all items in the group have the same ideal answer
-            ideal_count += consensus == ideal_answer_fn(group[0])
+            return consensus == ideal_answer_fn(group[0])
+        return 0
 
+    ideal_count = sum(await asyncio.gather(*itertools.starmap(index, groups.items())))
     return grouped_consensus, ideal_count / len(groups) if groups else 0.0
 
 
+async def evaluate_consensus(
+    data: Sequence[TData],
+    extract_answer_fn: Callable[[TData], TAnswer | Awaitable[TAnswer]],
+    num_samples: int,
+    seed: np.random.Generator | random.Random | int | None = None,
+    consensus_callback: Callable[[TAnswer, int, int], Any] | None = None,
+) -> tuple[list[tuple[TAnswer, int]], TAnswer]:
+    """
+    Create consensus bins given data.
+
+    Args:
+        data: Data to evaluate consensus upon, length is called N.
+        extract_answer_fn: Function to extract the actual answer from a datum. It can
+            be async so this can be done using a LLM call.
+        num_samples: Number of samples to choose from the N total.
+        seed: Optional seed for sampling.
+        consensus_callback: Optional callback function called just after computing
+            consensus, it's passed the consensus answer, consensus size, and
+            sample size. This is useful for stuff like summary statistics.
+
+    Returns:
+        Two-tuple of consensus bins generated by collections.Counter.most_common and the
+            consensus answer.
+    """
+    rand = (
+        seed
+        if isinstance(seed, np.random.Generator | random.Random)
+        else np.random.default_rng(seed)
+    )
+
+    if len(data) < num_samples:  # Too few items, sample with replacement
+        sampled: Iterable[TData] = [rand.choice(data) for _ in range(num_samples)]  # type: ignore[arg-type]
+    elif isinstance(rand, random.Random):  # Built-in random sample with replacement
+        sampled = rand.sample(data, num_samples)
```

Comment:
I don't think we should be supporting sampling with replacement at all. For one, the standard error you have above assumes IID trials. And this will increase the variance of the consensus answer.

What motivated supporting it even if `len(data) >= num_samples`?

Reply:
Yeah this comment should have said `Built-in random sample without replacement` as `random.Random.sample ` is without replacement. But due to a copy-paste error I introduced during this PR it said `with replacement`, which I have now fixed. Sorry for confusion here